### ISSUE TITLE
HIRAM issue with Release 20241227

### ISSUE THREAD

**Issue by ruenes1** [MiSTer-devel/ao486_MiSTer#177]
HIRAM does not seem to work in the latest release (20241227). 

For example, the top300 image with the previous core release in HIRAM mode had  627K of base memory:

![old](https://github.com/user-attachments/assets/50566b70-105f-4d1b-9deb-bdac7186e8ea)

The exact same image and settings with the latest core release results in only 529K, with nothing being loaded into upper mem:

![new](https://github.com/user-attachments/assets/d31beed4-b097-4556-8d1d-78aadb3e7ff2)

This prevents many games to load. EMM386 and QEMM memory modes seem to work fine. Only HIRAM seems affected. At least two users have confirmed this issue with different DOS versions (6.2.2 and Win98 embedded DOS)

Could this be related to the recent commit  #169 ?

**Comment by sorgelig** [comment]
Most last changes are came from @humply
So may be he can comment on this issue

**Comment by humply** [comment]
Yes, this is exactly the expected result of commit #169 that fixed UMA detection.

If you ever tried to run MS-DOS 6.22 with plain `HIMEM.SYS` and `EMM386.EXE` (without any memory range options) on the previous AO486 core, EMM386 would show the following error:
- Unable to set page frame base address

EMM386 wants to use UMA for the page frame but could not find any free space. The reason for this is because it detects the UMA as RAM (non-free). Memory drivers basically check for writable UMA to detect RAM there. The MS-DOS 6.22 tool `MSD` gives a clear overview of the UMA and what's detected there.

HiRAM binds available RAM areas (auxiliary memory cards, shadow-RAM, etc.) between the segments A000h-F000h into upper memory. So on the previous core you could indeed see HiRAM detect D000h-EFFFh as RAM. That's why you saw 128 kb of upper memory being created by HiRAM.

The most common configuration however is without RAM in the UMA (like it is now). Other emulators also default to this configuration.

https://www.vogonswiki.com/index.php/DOS_memory_management
> The upper memory area (UMA) is memory in the range between 640 kb and 1 Mb. By default there is no RAM in this range as it is reserved for use with hardware that is able to map own memory to this range.

The most common way to use free UMA is through the creation of upper memory blocks (UMBs) instead of having RAM there (which was not commonly used at the time) and having to use HiRAM.

So the way it works now is probably the more common and most compatible configuration in terms of what games from that time would expect.

It is however possible to create a menu option and make the UMA regions D000h-DFFFh and/or E000h-EFFFh writable again so it acts as RAM, but I'm not sure what software would actually benefit from it compared to using UMBs.

**Comment by ruenes1** [comment]
Thanks for the detailed explanation (and your great contributions to the core!). This now makes sense and looks more standard.

The previous behaviour, however, was quite handy because it allowed to load a bunch of common drivers (mouse, cdrom, etc.) without going through (painful) mem optimization, or having to create per-game/app config files. But I'm not sure if this 'non-standard' way of managing mem had any negative consequences.

**Comment by humply** [comment]
You're welcome, I hope I explained it well enough. You can find more information on the subject online if you're interested.

Having looked at the code I'm quite sure the RAM in the UMA was there by accident in the previous cores. I can understand that being able to use HiRAM had some benefits in terms of the ease with which memory could be configured, but it was also kind of odd and certainly not the default for most 386 and 486 computers. I think RAM expansion cards were more of a thing on older systems (the HiRAM use cases in the documentation also hint at this, e.g. "8086/88 or 80286 without extended memory, but with auxiliary memory").

Although adding extra menu options for UMA RAM is possible, I think it's not needed and would needlessly complicate memory configurations even more. Besides this, I'm also not sure about any HiRAM compatibility issues.

So HiRAM still works fine (it correctly doesn't detect any RAM in UMA now) and you will have to do memory optimizations just like on a real 486. 😄

**Comment by sorgelig** [comment]
I cannot say for sure as many years passed, but i remember i made RAM is UMA specifically to have more memory for drivers and TRS in HIRAM.
